File: app/src/main/java/net/squanchy/search/view/GridSpanSizeLookup.java
package net.squanchy.search.view;

import android.support.v7.widget.GridLayoutManager;

import net.squanchy.search.view.SearchAdapter.ViewTypeId;

class GridSpanSizeLookup extends GridLayoutManager.SpanSizeLookup {

    private static final int SINGLE_COLUMN_SPAN_SIZE = 1;

    private final int columnCount;
    private final ItemsAdapter itemsAdapter;

    GridSpanSizeLookup(ItemsAdapter adapter, int columnCount) {
        this.itemsAdapter = adapter;
        this.columnCount = columnCount;

        super.setSpanIndexCacheEnabled(true);
    }

    @Override
    public int getSpanSize(int position) {
        if (itemsAdapter.isEmpty()) {
            return SINGLE_COLUMN_SPAN_SIZE;
        }

        return getSpanSizeFor(itemsAdapter.viewTypeAtAbsolutePosition(position));
    }

    private int getSpanSizeFor(@ViewTypeId int viewTypeId) {
        if (viewTypeId == ViewTypeId.HEADER || viewTypeId == ViewTypeId.EVENT) {
            return columnCount;
        } else {
            return SINGLE_COLUMN_SPAN_SIZE;
        }
    }
}

File: app/src/main/java/net/squanchy/search/view/GridSpanSizeLookup.kt
package net.squanchy.search.view

import androidx.recyclerview.widget.GridLayoutManager
import net.squanchy.search.SearchListElement

internal class GridSpanSizeLookup(
    private val itemRetriever: (Int) -> SearchListElement,
    private val isAdapterEmpty: () -> Boolean,
    private val columnCount: Int
) : GridLayoutManager.SpanSizeLookup() {

    init {
        super.setSpanIndexCacheEnabled(true)
    }

    override fun getSpanSize(position: Int): Int {
        if (isAdapterEmpty()) {
            return SINGLE_COLUMN_SPAN_SIZE
        } else {
            return getSpanSizeFor(itemRetriever(position))
        }
    }

    private fun getSpanSizeFor(element: SearchListElement): Int =
        when (element) {
            is SearchListElement.EventHeader -> columnCount
            is SearchListElement.SpeakerHeader -> columnCount
            is SearchListElement.EventElement -> columnCount
            is SearchListElement.SpeakerElement -> SINGLE_COLUMN_SPAN_SIZE
            is SearchListElement.AlgoliaLogo -> columnCount
        }

    companion object {
        private const val SINGLE_COLUMN_SPAN_SIZE = 1
    }
}
